docs/architecture/adr-001-umee-module.md

```diff
@@ -0,0 +1,90 @@
+# ADR 001: Umee Module
+
+## Changelog
+
+- April 28, 2021: Initial Draft (@alexanderbez)
+
+## Status
+
+Proposed
+
+## Context
+
+Umee is a Universal Capital Facility that can collateralize assets on one
+blockchain towards borrowing assets on another blockchain. The platform specializes
+in allowing staked assets from POS blockchains to be used as collateral for
+borrowing across blockchains. The platform uses a combination of algorithmically
+determined interest rates based on market driven conditions.
+
+For the initial MVP implementation of the Umee network, we require the ability
+for users to be able to send ATOM tokens to a dedicated pool in the Umee network
+via IBC. ATOM tokens deposited into the Umee pool will mint a derivative meToken
+in a one-to-one ratio, i.e. one ATOM mints one meToken derivative.
+
+Validators on the Umee network will take these deposited ATOM tokens and delegate
+them to a set of governance-controlled validators on the ATOM source chain,
+e.g. the Cosmos Hub.
+
+A user that sent ATOM tokens to the Umee network can then take their derivative
+meTokens and send them to Ethereum via a bridge where a synthetic ERC-20 version
+of the meTokens are minted. Once sent, the meTokens are locked in Umee and the
+user can then freely trade and operate within Ethereum's DeFi ecosystem with the
+synthetic ERC-20 meTokens.
+
+## Decision
+
+To support users sending ATOM tokens to the Umee network via IBC, we will create
+a `x/umee` module that contains a dedicated `ModuleAccount` that represents the
+ATOM pool. In addition, we will implement a [custom IBC module](https://github.com/cosmos/ibc-go/blob/v1.0.0-alpha1/docs/custom.md).
+The custom IBC module will handle receiving custom packets on a custom Umee port.
+The custom module will be nearly identical to the ICS-20 module, except that it
+will perform the following:
+
+- Automatically send source ATOM tokens to the dedicated `ModuleAccount` pool
+  address instead of a user supplied address.
+- Mark the transfer's amount in a pending queue, `PendingDelegationQueue`, to be
+  later used for delegation on the source chain.
+- Track the sender and amount in the module's state.
+- Mint derivative meTokens and send them to the source sender's Umee account.
+
+At the beginning of a staking epoch, E<sub>1</sub> in `BeginBlock`, the `x/umee`
+module will perform the following:
+
+- Move the funds from the `PendingDelegationQueue` into another queue,
+  `PrevPendingDelegationQueue`. This queue is locked as funds cannot be sent to
+  it. In addition, these are the funds that will be delegated to the source chain
+  during the current staking epoch.
+- Clear out `PendingDelegationQueue`.
+
+The validators in the current staking epoch, E<sub>1</sub>, will then construct,
+sign, and broadcast multisig transactions to the Umee network. Once enough
```

Comment:
what's the multisig threshold? what happens if a validator doesn't sign in the current epoch? are there any penalties / rewards for uptime in the signing process?

Reply:
1/2. That's a good question. The multsig threshold will have to be power-based, but since we cannot actually construct a multsig like this, what we can do instead is the make the threshold 1/n. Then, on chain, we ensure 2/3 have signed. As for how and when it's created, each validator will need an account on the Hub with key `K` and they will use the same key `K` on umee for the multisig. When they create their validator on Umee, they will provide the account associated with `K`. This is just a rough idea.

3. Validators are changed each epoch, so if a validator wants to join, they'll need enough stake to enter the set in the _next_ epoch.

4. I haven't thought about this. We either do nothing or penalize them somehow via a slash or something. Or perhaps we keep a rolling window and slash if they don't sign "enough".

5. Rewards? No. Penalties? Perhaps. @brentxu have you thought about this?

Reply:
There was an ADR for weighted multisig a while ago: https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-019-multisigs.md

This was back when the multisig was part of Tendermint.

Perhaps it makes sense to extend the following:
https://github.com/cosmos/cosmos-sdk/blob/master/crypto/types/multisig/multisignature.go

With the above design it seems like it would be possible to alter the weights of the multisig depending on staking power... However it would have to be deployed on the Hub and all other supported chains based on this design.

Reply:
> We either do nothing or penalize them somehow via a slash or something

I think this is a safe default. Rolling window approach could also make sense, it all depends how much slack you'd want to give to the validators. If I understand correctly that would mean signing should give rewards, not signing means no reward, if not signing long enough then slashing.

Reply:
Yes @khoslaventures that is the multi-sig we will use -- the standard multi-sig used by the SDK.

Reply:
Upon further thought, I think all validators will have to sign. We cannot do 1/n because that validator can simply just broadcast transactions themselves.